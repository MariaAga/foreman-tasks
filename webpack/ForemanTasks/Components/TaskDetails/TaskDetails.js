import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Tab, Tabs } from 'patternfly-react';
import Skeleton from 'react-loading-skeleton';
import { translate as __ } from 'foremanReact/common/I18n';
import { STATUS } from 'foremanReact/constants';
import Task from './Components/Task';
import RunningSteps from './Components/RunningSteps';
import Errors from './Components/Errors';
import Locks from './Components/Locks';
import Raw from './Components/Raw';
import { getTaskID } from './TasksDetailsHelper';

import './TaskDetails.scss';

class TaskDetails extends Component {
  componentDidMount() {
    const { timeoutId, refetchTaskDetails, fetchTaskDetails } = this.props;

    fetchTaskDetails(getTaskID(), timeoutId, refetchTaskDetails);
  }
  componentWillUnmount() {
    this.props.taskReloadStop(this.props.timeoutId);
  }
  taskProgressToggle = () => {
    const {
      timeoutId,
      refetchTaskDetails,
      id,
      loading,
      taskReloadStop,
      taskReloadStart,
    } = this.props;
    if (timeoutId) {
      taskReloadStop(timeoutId);
    } else {
      taskReloadStart(timeoutId, refetchTaskDetails, id, loading);
    }
  };

  render() {
    const {
      externalId,
      startedAt,
      endedAt,
      label,
      input,
      output,
      executionPlan,
      failedSteps,
      runningSteps,
      locks,
      cancelStep,
      status,
      isData,
    } = this.props;
    const id = getTaskID();
    const resumable = executionPlan ? executionPlan.state === 'paused' : false;
    const cancellable = executionPlan ? executionPlan.cancellable : false;
    const loading = status === STATUS.PENDING && !isData;
    return (
      <div className="task-details-react well">
        <Tabs defaultActiveKey={1} animation={false} id="task-details-tabs">
          <Tab eventKey={1} title={__('Task')}>
            {loading ? (
              <Skeleton height={350} />
            ) : (
              <Task
                {...{
                  ...this.props,
                  cancellable,
                  resumable,
                  id,
                  status,
                  taskProgressToggle: this.taskProgressToggle,
                }}
              />
            )}
          </Tab>
          <Tab eventKey={2} disabled={loading} title={__('Running Steps')}>
            <RunningSteps
              runningSteps={runningSteps}
              id={id}
              cancelStep={cancelStep}
              taskReload={this.props.taskReload}
              taskProgressToggle={this.taskProgressToggle}
            />
          </Tab>
          <Tab eventKey={3} disabled={loading} title={__('Errors')}>
            <Errors executionPlan={executionPlan} failedSteps={failedSteps} />
          </Tab>
          <Tab eventKey={4} disabled={loading} title={__('Locks')}>
            <Locks locks={locks} />
          </Tab>
          <Tab eventKey={5} disabled={loading} title={__('Raw')}>
            <Raw
              {...{ id, label, startedAt, endedAt, input, output, externalId }}
            />
          </Tab>
        </Tabs>
      </div>
    );
  }
}

TaskDetails.propTypes = {
  label: PropTypes.string,
  fetchTaskDetails: PropTypes.func.isRequired,
  runningSteps: PropTypes.array,
  cancelStep: PropTypes.func.isRequired,
  taskReload: PropTypes.bool.isRequired,
  ...Task.propTypes,
  ...Errors.propTypes,
  ...Locks.propTypes,
  ...Raw.propTypes,
};
TaskDetails.defaultProps = {
  label: '',
  runningSteps: [],
  ...Task.defaultProps,
  ...RunningSteps.defaultProps,
  ...Errors.defaultProps,
  ...Locks.defaultProps,
  ...Raw.defaultProps,
};

export default TaskDetails;
