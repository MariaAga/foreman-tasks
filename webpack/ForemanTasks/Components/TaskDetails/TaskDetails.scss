.task-details-react {
  .progress-label-top-right {
    font-size: 11px;
    text-align: right;
  }
  a,
  button {
    margin-right: 3px;
  }
  .container {
    margin: 0;
  }
  .spin {
    -webkit-animation: spin 1s infinite linear;
    -moz-animation: spin 1s infinite linear;
    -o-animation: spin 1s infinite linear;
    animation: spin 1s infinite linear;
    -webkit-transform-origin: 50% 50%;
    transform-origin: 50% 50%;
    -ms-transform-origin: 50% 50%; /* IE 9 */
  }

  @-moz-keyframes spin {
    from {
      -moz-transform: rotate(0deg);
    }
    to {
      -moz-transform: rotate(360deg);
    }
  }

  @-webkit-keyframes spin {
    from {
      -webkit-transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(360deg);
    }
  }

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
  .dynflow-button > span {
    pointer-events: auto;
  }

  pre {
    white-space: pre;
    word-break: normal;
  }

  .param-name {
    display: inline-block;
    width: 10em;
  }
}
