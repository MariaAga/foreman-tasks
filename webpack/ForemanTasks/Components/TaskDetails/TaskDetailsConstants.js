export const FOREMAN_TASK_DETAILS = 'FOREMAN_TASK_DETAILS';
export const FOREMAN_TASK_DETAILS_SUCCESS = 'FOREMAN_TASK_DETAILS_SUCCESS';

export const TASK_STEP_CANCEL = 'TASK_STEP_CANCEL';
