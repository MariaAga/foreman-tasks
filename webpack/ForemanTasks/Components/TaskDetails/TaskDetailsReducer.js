import Immutable from 'seamless-immutable';
import {
  FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  FOREMAN_TASK_DETAILS_STOP_PULLING,
  FOREMAN_TASK_DETAILS_START_PULLING,
  FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
  FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
} from './TaskDetailsConstants';

const initialState = Immutable({});

export default (state = initialState, action) => {
  const { type, payload } = action;

  switch (type) {
    case FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS:
      return state.merge({
        ...payload,
      });
    case FOREMAN_TASK_DETAILS_STOP_PULLING:
      if (payload.timeoutId) {
        clearTimeout(payload.timeoutId);
      }
      return state.merge({ taskReload: false, timeoutId: null });
    case FOREMAN_TASK_DETAILS_START_PULLING:
      if (!payload.timeoutId) {
        payload.timeoutId = setInterval(
          () => payload.refetchTaskDetails(payload.id),
          5000
        );
      }
      return state.merge({ taskReload: true, timeoutId: payload.timeoutId });
    case FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL:
      return state.set('showUnlockModal', !state.showUnlockModal);
    case FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL:
      return state.set('showForceUnlockModal', !state.showForceUnlockModal);
    default:
      return state;
  }
};
