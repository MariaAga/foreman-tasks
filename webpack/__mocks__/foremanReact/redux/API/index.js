export const API = {
  get: jest.fn(),
  put: jest.fn(),
  post: jest.fn(),
  delete: jest.fn(),
  patch: jest.fn(),
};

export const get = data => ({ type: 'some-type', ...data });
