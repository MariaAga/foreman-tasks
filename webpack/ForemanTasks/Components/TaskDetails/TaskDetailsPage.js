import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import PageLayout from 'foremanReact/pages/common/PageLayout/PageLayout';
import TaskDetails from './TaskDetails';
import { getTaskID } from './TasksDetailsHelper';

export const indexBreadcrumbItem = {
  caption: __('Tasks'),
  url: '/foreman_tasks/tasks',
};

export const breadcrumbTaskItem = task => ({
  caption: task.action || __('N/A'),
  url: `/foreman_tasks/tasks/${getTaskID()}/details`,
});

class TaskDetailsPage extends Component {
  componentDidMount() {
    const { timeoutId, refetchTaskDetails, fetchTaskDetails } = this.props;

    fetchTaskDetails(getTaskID(), timeoutId, refetchTaskDetails);
  }
  componentWillUnmount() {
    this.props.taskReloadStop(this.props.timeoutId);
  }
  render() {
    return (
      <PageLayout
        searchable={false}
        breadcrumbOptions={{
          isSwitchable: true,
          breadcrumbItems: [
            indexBreadcrumbItem,
            breadcrumbTaskItem(this.props),
          ],
          resource: {
            nameField: 'action',
            resourceUrl: '/foreman_tasks/api/tasks',
            switcherItemUrl: '/foreman_tasks/tasks/:id',
          },
        }}
      >
        <TaskDetails {...this.props} />
      </PageLayout>
    );
  }
}

TaskDetailsPage.propTypes = {
  fetchTaskDetails: PropTypes.func,
  refetchTaskDetails: PropTypes.func,
  taskReloadStop: PropTypes.func,
  timeoutId: PropTypes.number,
};

TaskDetailsPage.defaultProps = {
  fetchTaskDetails: () => null,
  refetchTaskDetails: () => null,
  taskReloadStop: () => null,
  timeoutId: null,
};
export default TaskDetailsPage;
