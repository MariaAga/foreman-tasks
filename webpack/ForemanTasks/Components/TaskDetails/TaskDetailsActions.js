import { API } from 'foremanReact/redux/API';
import {
  showLoading,
  hideLoading,
} from 'foremanReact/components/Layout/LayoutActions';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { translate as __ } from 'foremanReact/common/I18n';
import {
  FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST,
  FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  FOREMAN_TASK_DETAILS_FETCH_TASK_FAILURE,
  FOREMAN_TASK_DETAILS_STOP_POLLING,
  FOREMAN_TASK_DETAILS_START_POLLING,
  TASK_STEP_CANCEL_REQUEST,
  TASK_STEP_CANCEL_FAILURE,
  TASK_STEP_CANCEL_SUCCESS,
} from './TaskDetailsConstants';
import {
  errorToastData,
  infoToastData,
  successToastData,
} from '../common/ToastsHelpers';

export const taskReloadStop = timeoutId => {
  if (timeoutId) {
    clearTimeout(timeoutId);
  }
  return {
    type: FOREMAN_TASK_DETAILS_STOP_POLLING,
    payload: { timeoutId },
  };
};

export const taskReloadStart = (
  timeoutId,
  refetchTaskDetails,
  id,
  loading = false
) => {
  if (!timeoutId) {
    timeoutId = setInterval(
      () => refetchTaskDetails(id, loading, timeoutId),
      5000
    );
  }
  return {
    type: FOREMAN_TASK_DETAILS_START_POLLING,
    payload: { timeoutId },
  };
};

export const refetchTaskDetails = (id, loading, timeoutId) => dispatch => {
  if (!loading) {
    showLoading();
    dispatch(startRequest());
    reloadTasksDetails(id, timeoutId, dispatch);
  }
};

const reloadTasksDetails = async (id, timeoutId, dispatch) => {
  try {
    const { data } = await API.get(
      `/foreman_tasks/api/tasks/${id}/details?include_permissions`
    );
    dispatch(requestSuccess(data));
  } catch (error) {
    dispatch(taskReloadStop(timeoutId));
    dispatch(requestFailure(error));
  } finally {
    hideLoading();
  }
};

export const fetchTaskDetails = (
  id,
  timeoutId,
  refetchTaskDetailsAction
) => async dispatch => {
  showLoading();
  dispatch(startRequest());
  await getTasksDetails(id, dispatch, timeoutId, refetchTaskDetailsAction);
};

const getTasksDetails = async (
  id,
  dispatch,
  timeoutId,
  refetchTaskDetailsAction
) => {
  try {
    const { data } = await API.get(
      `/foreman_tasks/api/tasks/${id}/details?include_permissions`
    );
    dispatch(requestSuccess(data));
    if (data.state !== 'stopped') {
      dispatch(taskReloadStart(timeoutId, refetchTaskDetailsAction, id));
    }
  } catch (error) {
    dispatch(requestFailure(error));
    if (timeoutId) dispatch(taskReloadStop(timeoutId));
  } finally {
    hideLoading();
  }
};

const startRequest = () => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST,
});

const requestSuccess = data => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS,
  payload: data,
});

const requestFailure = error => ({
  type: FOREMAN_TASK_DETAILS_FETCH_TASK_FAILURE,
  payload: error,
});

export const cancelStep = (taskId, stepId) => async dispatch => {
  dispatch({ type: TASK_STEP_CANCEL_REQUEST });
  dispatch(addToast(infoToastData(`${__('Trying to cancel step')} ${stepId}`)));
  try {
    await API.post(
      `/foreman_tasks/tasks/${taskId}/cancel_step?step_id=${stepId}`
    );
    dispatch({ type: TASK_STEP_CANCEL_SUCCESS });
    dispatch(addToast(successToastData(`${stepId} {__('Step Canceled')}`)));
  } catch (error) {
    dispatch({ type: TASK_STEP_CANCEL_FAILURE, payload: error });
    dispatch(
      addToast(
        errorToastData(
          `${__('Could not cancel step.')} ${__(
            'Error:'
          )} ${stepId} ${error.response &&
            error.response.data &&
            error.response.data.error}`
        )
      )
    );
  }
};
