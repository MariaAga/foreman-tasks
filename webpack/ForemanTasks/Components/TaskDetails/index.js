import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import TaskDetails from './TaskDetails';
import * as taskDetailsActions from './TaskDetailsActions';
import * as taskActions from '../TaskActions';
import reducer from './TaskDetailsReducer';
import {
  selectEndedAt,
  selectStartAt,
  selectStartBefore,
  selectStartedAt,
  selectLocks,
  selectInput,
  selectOutput,
  selectResumable,
  selectCancellable,
  selectHelp,
  selectErrors,
  selectProgress,
  selectUsername,
  selectLabel,
  selectExecutionPlan,
  selectFailedSteps,
  selectRunningSteps,
  selectHasSubTasks,
  selectUsernamePath,
  selectAction,
  selectState,
  selectResult,
  selectTimeoutId,
  selectTaskReload,
  selectParentTask,
  selectExternalId,
  selectDynflowEnableConsole,
  selectCanEdit,
  selectStatus,
  selectError,
  selectIsData,
} from './TaskDetailsSelectors';

const mapStateToProps = state => ({
  startAt: selectStartAt(state),
  startBefore: selectStartBefore(state),
  startedAt: selectStartedAt(state),
  endedAt: selectEndedAt(state),
  input: selectInput(state),
  output: selectOutput(state),
  resumable: selectResumable(state),
  cancellable: selectCancellable(state),
  errors: selectErrors(state),
  progress: selectProgress(state),
  username: selectUsername(state),
  label: selectLabel(state),
  executionPlan: selectExecutionPlan(state),
  failedSteps: selectFailedSteps(state),
  runningSteps: selectRunningSteps(state),
  help: selectHelp(state),
  hasSubTasks: selectHasSubTasks(state),
  locks: selectLocks(state),
  usernamePath: selectUsernamePath(state),
  action: selectAction(state),
  state: selectState(state),
  result: selectResult(state),
  timeoutId: selectTimeoutId(state),
  taskReload: selectTaskReload(state),
  parentTask: selectParentTask(state),
  externalId: selectExternalId(state),
  dynflowEnableConsole: selectDynflowEnableConsole(state),
  canEdit: selectCanEdit(state),
  status: selectStatus(state),
  error: selectError(state),
  isData: selectIsData(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ ...taskActions, ...taskDetailsActions }, dispatch);

export const reducers = { taskDetails: reducer };

export default connect(mapStateToProps, mapDispatchToProps)(TaskDetails);
