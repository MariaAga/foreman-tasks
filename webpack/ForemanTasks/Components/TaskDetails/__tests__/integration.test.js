import React from 'react';

import { API } from 'foremanReact/redux/API';
import { IntegrationTestHelper } from '@theforeman/test';
import { STATUS } from 'foremanReact/constants';

import TaskDetails, { reducers } from '../index';
import { selectForemanTasks } from '../../../ForemanTasksSelectors';
import { minProps } from './TaskDetails.fixtures';

jest.mock('../../../ForemanTasksSelectors');
jest.mock('foremanReact/redux/API');
selectForemanTasks.mockImplementation(state => state);
describe('TaskDetails integration test', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    delete window.location;
    window.location = new URL(
      'https://foreman.com/foreman_tasks/tasks/a15dd820-32f1-4ced-9ab7-c0fab8234c47/'
    );
  });
  it('should flow', async () => {
    window.location.reload = jest.fn();
    API.get.mockImplementation(() => ({
      data: { label: 'new-label', state: 'stopped' },
    }));

    const integrationTestHelper = new IntegrationTestHelper(reducers);

    const component = integrationTestHelper.mount(
      <TaskDetails status={STATUS.RESOLVED} {...minProps} id="test" />
    );
    integrationTestHelper.takeStoreSnapshot('initial state');
    await IntegrationTestHelper.flushAllPromises();
    component.update();
    const reloadButton = component.find('.reload-button').at(0);
    reloadButton.simulate('click');
    await IntegrationTestHelper.flushAllPromises();

    integrationTestHelper.takeActionsSnapshot('task details reload start');

    reloadButton.simulate('click');
    await IntegrationTestHelper.flushAllPromises();
    jest.advanceTimersByTime(1);
    integrationTestHelper.takeActionsSnapshot(
      'task details reload stop after reload button click'
    );
    jest.useRealTimers();
  });

  it('should change the data after reload', async () => {
    API.get.mockImplementationOnce(() => ({
      data: { label: 'new-label-unchanged', state: 'stopped' },
    }));
    const integrationTestHelper = new IntegrationTestHelper(reducers);

    const component = integrationTestHelper.mount(
      <TaskDetails {...minProps} id="test" />
    );

    await IntegrationTestHelper.flushAllPromises();
    component.update();

    API.get.mockImplementationOnce(() => ({
      data: { label: 'new-label-changed', state: 'stopped' },
    }));
    integrationTestHelper.takeStoreSnapshot('initial state');
    const reloadButton = component.find('.reload-button').at(0);
    reloadButton.simulate('click');

    jest.runOnlyPendingTimers();
    await IntegrationTestHelper.flushAllPromises();
    integrationTestHelper.takeStoreSnapshot('task details data should change');
    jest.useRealTimers();
  });
});
