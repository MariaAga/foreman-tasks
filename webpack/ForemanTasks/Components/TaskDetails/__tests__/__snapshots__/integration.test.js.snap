// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDetails integration test should change the data after reload: initial state 1`] = `
Object {
  "taskDetails": Object {
    "isData": true,
    "label": "new-label-unchanged",
    "state": "stopped",
    "status": "RESOLVED",
    "taskReload": false,
    "timeoutId": null,
  },
}
`;

exports[`TaskDetails integration test should change the data after reload: task details data should change 1`] = `
Object {
  "taskDetails": Object {
    "isData": true,
    "label": "new-label-changed",
    "state": "stopped",
    "status": "RESOLVED",
    "taskReload": false,
    "timeoutId": null,
  },
}
`;

exports[`TaskDetails integration test should flow: initial state 1`] = `
Object {
  "taskDetails": Object {
    "isData": false,
    "status": "PENDING",
  },
}
`;

exports[`TaskDetails integration test should flow: task details reload start 1`] = `
Array [
  Array [
    Object {
      "type": "FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "label": "new-label",
        "state": "stopped",
      },
      "type": "FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "timeoutId": 3,
      },
      "type": "FOREMAN_TASK_DETAILS_START_POLLING",
    },
  ],
]
`;

exports[`TaskDetails integration test should flow: task details reload stop after reload button click 1`] = `
Array [
  Array [
    Object {
      "type": "FOREMAN_TASK_DETAILS_FETCH_TASK_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "label": "new-label",
        "state": "stopped",
      },
      "type": "FOREMAN_TASK_DETAILS_FETCH_TASK_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "timeoutId": 3,
      },
      "type": "FOREMAN_TASK_DETAILS_START_POLLING",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "timeoutId": 3,
      },
      "type": "FOREMAN_TASK_DETAILS_STOP_POLLING",
    },
  ],
]
`;
